lib/graphs/digraph.nit

```diff
@@ -0,0 +1,606 @@
+	# assert g.num_vertices == 2
+	# ~~~
+	fun add_vertex(u: V) is abstract
+
+	# Returns true if and only if `u` exists in this graph.
+	#
+	# ~~~
+	# var g = new HashMapDigraph[Int]
+	# g.add_vertex(1)
+	# assert g.has_vertex(1)
+	# assert not g.has_vertex(0)
+	# g.add_vertex(1)
+	# assert g.has_vertex(1)
+	# assert not g.has_vertex(0)
+	# ~~~
+	fun has_vertex(u: V): Bool is abstract
+
+	# Removes the vertex `u` from this graph and all its incident arcs.
+	#
+	# If the vertex does not exist in the graph, then nothing happens.
+	#
+	# ~~~
+	# var g = new HashMapDigraph[Int]
+	# g.add_vertex(0)
+	# g.add_vertex(1)
+	# assert g.has_vertex(0)
+	# g.remove_vertex(0)
+	# assert not g.has_vertex(0)
+	# ~~~
+	fun remove_vertex(u: V) is abstract
+
+	# Adds the arc `(u,v)` to this graph.
+	#
+	# If the arc already exists in the graph, then nothing happens.
+	# If vertex `u` or vertex `v` do not exist in the graph, they are added.
+	fun add_arc(u, v: V) is abstract
+
+	# Returns true if and only if `(u,v)` is an arc in this graph.
+	fun has_arc(u, v: V): Bool is abstract
+
+	# Removes the arc `(u,v)` from this graph.
+	#
+	# If the arc does not exist in the graph, then nothing happens.
+	fun remove_arc(u, v: V) is abstract
+
+	# Returns the predecessors of `u`.
+	#
+	# If `u` does not exist, then an empty collection is returned.
+	fun predecessors(u: V): Collection[V] is abstract
+
+	# Returns the successors of `u`.
+	#
+	# If `u` does not exist, then an empty collection is returned.
+	fun successors(u: V): Collection[V] is abstract
+
+	# Returns the vertices of this graph.
+	fun vertices: Collection[V] is abstract
+
+	# Returns the arcs of this graph.
+	fun arcs: Collection[Arc[V]] is abstract
+
+	## -------------------- ##
+	## Non abstract methods ##
+	## -------------------- ##
+
+	## ------------ ##
+	## Neighborhood ##
+	## ------------ ##
+
+	# Returns true if and only if `u` is a predecessor of `v`.
+	fun is_predecessor(u, v: V): Bool
+	do
+		return has_arc(u, v)
+	end
+
+	# Returns true if and only if `u` is a successor of `v`.
+	fun is_successor(u, v: V): Bool
+	do
+		return has_arc(v, u)
+	end
+
+	# Returns the incoming arcs of vertex `u`.
+	#
+	# If `u` is not in this graph, an empty array is returned.
+	fun incoming_arcs(u: V): Array[Arc[V]]
+	do
+		var arcs = new Array[Arc[V]]
+		for v in successors(u) do
+			arcs.add(new Arc[V](u, v))
+		end
+		return arcs
+	end
+
+	# Returns the outgoing arcs of vertex `u`.
+	#
+	# If `u` is not in this graph, an empty array is returned.
+	fun outgoing_arcs(u: V): Array[Arc[V]]
+	do
+		var arcs = new Array[Arc[V]]
+		for v in predecessors(u) do
+			arcs.add(new Arc[V](v, u))
+		end
+		return arcs
+	end
+
+	## ---------------------- ##
+	## String representations ##
+	## ---------------------- ##
+
+	redef fun to_s: String
+	do
+		var s = "vertices"
+		var t = "arcs"
+		if num_vertices <= 1 then s = "vertex"
+		if num_arcs <= 1 then t = "arc"
+		s = "Digraph of {num_vertices} {s} and {num_arcs} {t}\n"
```

Comment:
I am not a fan of having newlines in `to_s` because it may produce bad results when combined by clients that call `to_s` in an abstract way. It is just a gut feeling so I wont complain more.